10_rest/runnerdave/cmd/puppy_server/main.go

```diff
@@ -0,0 +1,112 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	puppy "github.com/anz-bank/go-course/10_rest/runnerdave/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/runnerdave/pkg/puppy/store"
+
+	"github.com/go-chi/chi"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	args           = os.Args[1:]
+	data           = kingpin.Flag("data", "data file").Short('d').Default("puppydata/data.json").ExistingFile()
+	port           = kingpin.Flag("port", "PORT").Short('p').Default("8080").String()
+	db             = kingpin.Flag("store", "STORE").Short('s').Default("sync").String()
+	out  io.Writer = os.Stdout
+)
+
+func main() {
+	_, err := kingpin.CommandLine.Parse(args)
+	if err != nil {
+		fmt.Fprintf(out, "Command line could not be parsed, error: %v", err)
+		return
+	}
+
+	d, _ := ioutil.ReadFile(*data)
+	puppies, uerr := unmarshalPuppies(d)
+	if uerr != nil {
+		return
+	}
+
+	s := loadPuppies(puppies, *db)
+	r := chi.NewRouter()
+
+	rs := puppy.RestStorer{Db: s}
+	r.Get(rs.GetPuppyRoute(), rs.GetPuppy)
+	r.Post(rs.PostPuppyRoute(), rs.CreatePuppy)
+	r.Put(rs.PutPuppyRoute(), rs.UpdatePuppy)
+	r.Delete(rs.DeletePuppyRoute(), rs.DeletePuppy)
+
+	portValue := fmt.Sprintf(":%v", *port)
+	serv := http.Server{Addr: portValue, Handler: r}
+	serr := serv.ListenAndServe()
+	if serr != http.ErrServerClosed {
+		fmt.Fprintf(out, "Could not start puppy server: %v", serr)
+	}
+}
+
+func loadPuppies(puppies []puppy.Puppy, db string) puppy.Storer {
+	var s puppy.Storer
+	switch db {
+	case "map":
+		mapStore, err := storeInMap(puppies)
+		if err != nil {
+			fmt.Fprintf(out, "Could not setup map storage:%v", err)
+		}
+		fmt.Fprintf(out, "Map store of puppies:%v", mapStore)
+		s = &mapStore
+	case "sync":
+		syncStore, err := storeInSync(puppies)
+		if err != nil {
+			fmt.Fprintf(out, "Could not setup sync storage:%v", err)
+		}
+		fmt.Fprintf(out, "Sync store of puppies:%v", syncStore)
+		s = syncStore
+	default:
+		fmt.Fprintf(out, "Invalid storage")
+	}
+	return s
+}
+
+func unmarshalPuppies(d []byte) ([]puppy.Puppy, error) {
+	puppies := []puppy.Puppy{}
+	if err := json.Unmarshal(d, &puppies); err != nil {
+		fmt.Fprintf(out, "Could not unmarshall puppies, error: %v", err)
```

Comment:
I'd get rid of all the debug messages and just print proper error handling in main.
Alternatively you can wrap your error:
`return nil, fmt.Errorf("cannot unmarshal puppies, error: %v", err)`

same below.

Reply:
done, no debug message in functions, only handling of errors in main